docs(aio): add Metadata doc

kapunahelewong opened the pull request:
## PR Checklist
Does please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?
Adds Chuck's metadata doc.

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Currently there is no metadata as it pertains to AOT.

Issue Number: N/A


## What is the new behavior?
Adds a doc explaining AOT and metadata. 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

kapunahelewong commented:
@chuckjaz I've made your edits but would love to hear any more thoughts and comments you may have. Thank you!

mary-poppins commented:
The angular.io preview for c302388d23fffec90b96173e91a2d604ae23c0a2 is available [here][1].

[1]: https://pr17712-c302388d23fffec90b96173e91a2d604ae23c0a2.ngbuilds.io/

wardbell commented:
The CI fail is due to low PWA score and is not related to this doc.

mary-poppins commented:
The angular.io preview for 2770b0a8fddfb31f65a6e48fa3cc6bca28fb7a5c is available [here][1].

[1]: https://pr17712-2770b0a8fddfb31f65a6e48fa3cc6bca28fb7a5c.ngbuilds.io/

mary-poppins commented:
You can preview 1ebd9a50a50208fa69f8557362aa2bca2504503a at https://pr17712-1ebd9a50a50208fa69f8557362aa2bca2504503a.ngbuilds.io/.

mary-poppins commented:
You can preview 387910a at https://pr17712-387910a.ngbuilds.io/.

mary-poppins commented:
You can preview e53dfaa at https://pr17712-e53dfaa.ngbuilds.io/.

mary-poppins commented:
You can preview 13744fd at https://pr17712-13744fd.ngbuilds.io/.

mary-poppins commented:
You can preview 4e95724 at https://pr17712-4e95724.ngbuilds.io/.

mary-poppins commented:
You can preview 02979b7 at https://pr17712-02979b7.ngbuilds.io/.

chuckjaz commented on aio/content/guide/metadata-errors.md:
```diff
@@ -0,0 +1,91 @@
+## Introduction
+
+This page lists errors Angular returns regarding metadata.
+<!--Chuck, could you help me use the right terminology here? Would it 
```
These are errors produced by `ngc` (or most likely invoked via  `ng build --aot`) so I would refer to them and Angular compiler errors.

kapunahelewong replied:
Updated.

chuckjaz commented on aio/content/guide/metadata-errors.md:
```diff
@@ -0,0 +1,91 @@
+or when you use an expression to specify the type parameters 
+of a type used in a class constructor.
+
+<!--What is a heritage clause?-->
```
We should call it the `extends` clause. It declares from which this class is derived. For example,

```ts
class SomeClass extends ParentClass {
   ...
}
```
In this case `extends ParentClass` is the `extends` clause.

kapunahelewong replied:
Thank you. Got it. Updated.

chuckjaz commented on aio/content/guide/metadata-errors.md:
```diff
@@ -0,0 +1,91 @@
+provide the initial value of the variable in an initializer clause.
+
+### `Reference to a non-exported class <class name>. Consider exporting the class`
+An annotation referenced a class in an annotation but the module didn't export it, for example, registered as a provider. 
```
Consider "used as a token in the providers array" or "used as the value of the `provide` field of a provider".

kapunahelewong replied:
Fixed.

chuckjaz commented on aio/content/guide/metadata-errors.md:
```diff
@@ -0,0 +1,91 @@
+
+### `Reference to a non-exported function`
+An annotation referenced a function that the module didn't export, for example, registered as a provider factory. 
+<!--Chuck, I have the same kind of question for "registered as a provider". Who is registering it 
```
Consider "used as the value of `useFactory` in a provider"

kapunahelewong replied:
Fixed.

mary-poppins commented:
You can preview f973f34 at https://pr17712-f973f34.ngbuilds.io/.

mary-poppins commented:
You can preview e3d8f5a at https://pr17712-e3d8f5a.ngbuilds.io/.

mary-poppins commented:
You can preview 0477057 at https://pr17712-0477057.ngbuilds.io/.

mary-poppins commented:
You can preview 04ac3b2 at https://pr17712-04ac3b2.ngbuilds.io/.

mary-poppins commented:
You can preview c2fe48e at https://pr17712-c2fe48e.ngbuilds.io/.

mary-poppins commented:
You can preview c5b4eb5 at https://pr17712-c5b4eb5.ngbuilds.io/.

mary-poppins commented:
You can preview 5c8eff4 at https://pr17712-5c8eff4.ngbuilds.io/.

mary-poppins commented:
You can preview 73149b4 at https://pr17712-73149b4.ngbuilds.io/.

mary-poppins commented:
You can preview 370a680 at https://pr17712-370a680.ngbuilds.io/.

mary-poppins commented:
You can preview c2e84f5 at https://pr17712-c2e84f5.ngbuilds.io/.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -1,118 +1,81 @@
 
-The `@Component` annotation records the `selector` and `template` for 
-the `TypicalComponent` using the [Reflect API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect). With the `selector` and `template` data in hand, the JIT compiler uses `Reflect` to produce the factory function.
+Instead of compiling the application when the web page loads, you can compile it with AOT by running the `ngc` stand-alone tool as part of your build process. The AOT compiler produces a number of files, including the self-sufficient, application JavaScript that ultimately runs in the browser.
```
Consider adding a reference to `ng build` which is the pirmary way people use `ngc`.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -1,118 +1,81 @@
 
-The `.metadata.json` does not try to interpret the meaning of the attribute, it
-just records a representation of the source; the `StaticReflector` handles the interpretion as well as places restrictions on what it interprets.
+It helps to think of the AOT compiler as having two phases: a code analysis phase and a code generation phase.
```
Consider making this statement stronger as it actually does have these to phases.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -303,7 +317,9 @@ Symbol            | Module
-Symbol            | Module
+The compiler only generates calls to these Angular functions.
+
+Function          | Module
```
The compiler does not generate calls to these functions but, rather, calls them and then records the result
of calling them. These functions generate the animation data that the animation engine uses at runtime. The compiler records the result of the functions instead of generating calls to them.

We might want to remove this as `@angular/animations` no longer uses them. They `@angular/core` versions her deprecated in 4.0 and `@angular/animations` introduces macro functions that are not a special case.

kapunahelewong replied:
Thank you for the clarification. Removed.

chuckjaz reviewed (approved):
LGTM save some changes you may wish to consider.

mary-poppins commented:
You can preview e92a462 at https://pr17712-e92a462.ngbuilds.io/.

mary-poppins commented:
You can preview acfd00a at https://pr17712-acfd00a.ngbuilds.io/.

juleskremer commented on aio/content/guide/metadata.md:
````diff
@@ -0,0 +1,486 @@
+}
+```
+In principle, Angular could interpret the metadata every time your app requests a new instance of `TypicalComponent`. 
+That would be horribly inefficient. 
````
Remove this paragraph; it is unnecessary.

kapunahelewong replied:
Done.

juleskremer commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+The JIT compiler has access to the runtime state of the application. When targeting the JIT compiler, you could write Angular metadata with the full power of JavaScript and reference dynamically created functions and values. You could even re-compile parts of the application on the fly.
+
+But JIT compilation has disadvantages. The browser must download the compiler library (it's big), parse the compiler itself, and then compile the application, all before the user can see or interact with the first screen. There is also some security risk in generating JavaScript code in the browser.
+
```
The JIT section can be removed; this document is about AOT & Metadata.

kapunahelewong replied:
Removed.

juleskremer commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+There's no longer a need for the compiler on the client. The browser downloads a smaller payload of safely-compiled, application module(s) and libraries that it can parse and run immediately.
+
+AOT compilation has limitations too. The compiler can't execute any part of the application in a browser nor does it have access to any runtime values. It must _statically analyze_ your source code to understand how the app is constructed and to interpret the Angular metadata without actually running the application.
+
```
Rework these paragraphs:  remove "instead of compiling" - instead introduce AOT.  Instead of "AOT has limitations" change the language to positive tone.

kapunahelewong replied:
Done.

juleskremer commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+
+The compiler's static analysis is clever but it can't comprehend everything that you can do runtime in full JavaScript. 
+Your Angular metadata object definitions must conform to the following general constraints:
+
```
Please remove tone:  no need for this, simply introduce the metadata restrictions without tone.

kapunahelewong replied:
Done.

juleskremer commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+
+Angular libraries are built with this option to ensure that all Angular `.metadata.json` files are clean.
+You may wish to do the same when building your own libraries.
+
```
Instead of "you may wish" -> indicate that this is a best practice.

kapunahelewong replied:
Done.

juleskremer commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+When the compiler later interprets this node, it reports an error that invites you to turn the arrow function into an _exported function_.
+
+You can cure the error by converting to this:
+
```
change cure to fix.

kapunahelewong replied:
Done.

kapunahelewong commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,486 @@
+That would be horribly inefficient. 
+
+Instead, the Angular compiler interprets the metadata _once_ and generates a _factory_ for `TypicalComponent`. When asked to create a `TypicalComponent` instance, Angular calls the factory, which produces a new visual element, bound to a new instance of the component class with its injected dependencies.
+
```
Done.

mary-poppins commented:
You can preview e5e9ee6 at https://pr17712-e5e9ee6.ngbuilds.io/.

kapunahelewong committed:
docs(aio): add Metadata guide based on Chuck’s docs

Chuck’s gist
https://gist.github.com/chuckjaz/65dcc2fd5f4f5463e492ed0cb93bca60#file-Angular%20Metadata-md
Also chuck’s doc on metadata-related errors (link needed)

mary-poppins commented:
You can preview 0ab5016 at https://pr17712-0ab5016.ngbuilds.io/.

mary-poppins commented:
You can preview f6e0d32 at https://pr17712-f6e0d32.ngbuilds.io/.

mary-poppins commented:
You can preview 166360d at https://pr17712-166360d.ngbuilds.io/.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Example(s)?
```
This is error is produced for the reference to `Window` in:

```ts
@Component({ })
export class F {
  constructor (public w: Window) {}
}
```

`Window` was not imported and refers to the ambient `Window` type which we cannot resolve as the template compiler only supports imported symbols.

wardbell replied:
Is the problem only in the signature? Could I do this?
```
@Component({ })
export class F {
  constructor (public w: any) {
    const win = w as Window;
  }
}
```

chuckjaz replied:
> Surely I have a way to refer to ambient types by some circumlocution.

You need to use `@Inject`. The restriction is that the ambient type cannot be a provider token as we cannot import it. However, you still can use it and `@Inject` a token with the type as implied by your example.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Are these statements and examples correct?
```
They look right to me. Note that this will be revised for v5 as the lambda cases will be automatically lowered.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: What is a "tagged template expression"?
```
An example of a tagged template expression (terminology collision with our template, unfortunately) is:

```ts
const expression = 'of a';
const rawString = String.raw`Some example ${expression} string`;
```

In this case `String.raw` is the tag.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Are new features really the primary cause?
```
Probably not. Unusual syntax is probably the more likely cause such as using `typeof A` or `{ [prop]: 'value' }` in metadata.

chuckjaz commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+Chuck: Are new features really the primary cause?
+Or do you get this error for _any expression_ that we don't support such as "spreads in literal arrays"?
+
+BTW, do we support classes built with [TypeScript mixins](https://www.typescriptlang.org/docs/handbook/mixins.html)?
```
Yes by ignoring the implicit error as noted above.

wardbell replied:
As noted above ("Symbol reference expected") I'm not quite sure when mix-in syntax is permitted and where it is not. I'll look for clarification up there.

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Example(s)? What do you _do_ about it?
```
I'm still stuck on this one:
- Is this NOT a problem for `Component` and `Directive` classes?
- is this (therefore) a problem for "service" classes (provided by DI)?
- what is the "symbol" that is expected and not found in this example? 
- does the error message identify the "symbol"?
- do I really need a mix-in example or can we illustrate with a more simple `extends`?
- what is the developer supposed to DO about this? how do they overcome it?

chuckjaz replied:
> the compiler ignores this error _only_ for `Component` and `Directive` classes?

Yes, but that they are the only classes it actually will be looking at that field so it is safe to say that this is always ignored in the extends clause.

> is this (therefore) a problem _only_ for pipes, modules, and "service" classes provided by DI?

No. We don't look at the ancestor for these types so we it is ignored here as well.

> what is the "symbol" that is expected and not found in this example?

It expected `Tagged(Person)` to be a single identifier. It found an expression instead.

> do I really need a mix-in example or can we illustrate with a more simple extends?

You need something that is valid TypeScript and is rejected by the Angular compiler. A mix-in is the simplest example.

> what is the developer supposed to DO about this? how do they overcome it?

A library author should be aware that an error symbol in this location is present in the `.metadata.json` but is ignored by the compiler.

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+// ERROR
+import { Foo, Bar } from './configuration';
+
+const foo: number;
```
Oops. Of course.

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Example(s)? How is this different from `Reference to a local (non-exported) symbol 'symbol name'`?
```
You used the phrase "this error" twice to refer to two different errors and I'm not sure which is which.

`export let SOME_VALUE: string;` is exported but not initialized.  Looking at your specs I think this produces the `'Variable not initialized'` error which translates to the `'Only initialized variables ...'` message. I can't explain why this should be an error if the variable is never used but set that aside.

>Btw, the test in which this error is produced concerns destructuring, not initialization. Do you want to break it into its own test?

I take it that the second problem is with `template: SOME_VALUE` where the problem is that, while TS knows how to reference `SOME_VALUE`, the AOT compiler could not reach it. Therefore, _this_  node gets the `Reference to a local symbol` error.

Do I understand correctly?

chuckjaz replied:
> You used the phrase "this error" twice 

I mean the error with the tag `only-initialized-variables`

> I take it that the second problem is with `template: SOME_VALUE` where the problem is that, while TS knows how to reference `SOME_VALUE`, the AOT compiler could not reach it. Therefore, _this_  node gets the `Reference to a local symbol` error.

No. I believe you are mixing up local symbols with uninitialized variables. `only-initialized-variables` is reported when the compiler needs the value of a variable but no value was provided. It happens independent of whether the variable is local or exported.  A `reference-to-a-local-symbol` will be produced when cannot fold the reference to a local symbol. For example, using a non-exported class as a provider token.

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+<div class="alert is-critical">
+
+Chuck: Here and elsewhere below you refer to "template compilation" or "the template compiler". 
+Are you talking about something other than the AOT compiler?
```
OK.  I simply refer to "the compiler".

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Is this example correct?
```
I stole the two examples from your specs that deliver this message.

wardbell commented on aio/content/guide/metadata.md:
```diff
@@ -0,0 +1,773 @@
+
+<div class="alert is-critical">
+
+Chuck: Example(s)?
```
I'll use your example.

wardbell reviewed (changes requested):
Thanks Chuck. The fifth commit reflects your recent feedback.

I still have a question about one of the errors (look for "chuck:") and would be pleased to have your confirmation/clarifications on the ones that I fixed.

Click [here](https://github.com/angular/angular/pull/17712#discussion_r132025495) to see your comment and my response as they have been hidden as "outdated" by my latest force push.

mary-poppins commented:
You can preview 17e8ef0 at https://pr17712-17e8ef0.ngbuilds.io/.

mary-poppins commented:
You can preview 3f2fb34 at https://pr17712-3f2fb34.ngbuilds.io/.

mary-poppins commented:
You can preview 2784640 at https://pr17712-2784640.ngbuilds.io/.

mary-poppins commented:
You can preview 3abd625 at https://pr17712-3abd625.ngbuilds.io/.

mary-poppins commented:
You can preview 459ff66 at https://pr17712-459ff66.ngbuilds.io/.

mary-poppins commented:
You can preview 41e0fb2 at https://pr17712-41e0fb2.ngbuilds.io/.

mary-poppins commented:
You can preview fb56d4c at https://pr17712-fb56d4c.ngbuilds.io/.

mary-poppins commented:
You can preview 2bcdb66 at https://pr17712-2bcdb66.ngbuilds.io/.

mary-poppins commented:
You can preview c945d8d at https://pr17712-c945d8d.ngbuilds.io/.

mary-poppins commented:
You can preview 02a2cf4 at https://pr17712-02a2cf4.ngbuilds.io/.

wardbell commented:
Chuck and I agreed that the explanation for the last error (`Symbol reference expected`) can wait for a future PR; it's better to get this published now as is ... which is a version he likes.

Time to merge it.
